README.md
<p align="center"><img src="https://habrastorage.org/webt/59/e2/71/59e271948a792190098780.png"></p>

[![GoDoc](https://godoc.org/github.com/furdarius/rabbitroutine?status.svg)](https://godoc.org/github.com/furdarius/rabbitroutine)
[![Build Status](https://travis-ci.org/furdarius/rabbitroutine.svg?branch=master)](https://travis-ci.org/furdarius/rabbitroutine)
[![Go Report Card](https://goreportcard.com/badge/github.com/furdarius/rabbitroutine)](https://goreportcard.com/report/github.com/furdarius/rabbitroutine)

# Rabbitmq Failover Routine

Small, user-friendly library for RabbitMQ with auto-reconnect and publish retry routine.
The library is designed to save the developer from the headache when working with RabbitMQ.

**rabbitroutine** solves your RabbitMQ reconnection problems:
* Handles connection errors and channels errors separately.
* Takes into account the need to [re-declare](https://godoc.org/github.com/furdarius/rabbitroutine#Consumer) entities in RabbitMQ after reconnection.
* Notifies of [errors](https://godoc.org/github.com/furdarius/rabbitroutine#Connector.AddAMQPNotifiedListener) and connection [retry attempts](https://godoc.org/github.com/furdarius/rabbitroutine#Connector.AddRetriedListener).
* Supports [FireAndForgetPublisher](https://godoc.org/github.com/furdarius/rabbitroutine#FireForgetPublisher) and [EnsurePublisher](https://godoc.org/github.com/furdarius/rabbitroutine#EnsurePublisher), that can be wrapped with [RetryPublisher](https://godoc.org/github.com/furdarius/rabbitroutine#RetryPublisher).
* Supports pool of channels used for publishing.
* Provides channels [pool size](https://godoc.org/github.com/furdarius/rabbitroutine#Pool.Size) statistics.

**Stop to do wrappers, do features!**

## Install
```
go get github.com/furdarius/rabbitroutine
```

### Adding as dependency by "go dep"
```
$ dep ensure -add github.com/furdarius/rabbitroutine
```

## Usage


### Consuming
You need to implement [Consumer](https://godoc.org/github.com/furdarius/rabbitroutine#Consumer) and register
it with [StartConsumer](https://godoc.org/github.com/furdarius/rabbitroutine#Connector.StartConsumer)
or with [StartMultipleConsumers](https://godoc.org/github.com/furdarius/rabbitroutine#Connector.StartMultipleConsumers).
When connection is established (*at first time or after reconnect*) `Declare` method is called. It can be used to
declare required RabbitMQ entities ([consumer example](https://github.com/furdarius/rabbitroutine/blob/master/consumer_example_test.go)). 


Usage example:

```go

// Consumer declares your own RabbitMQ consumer implementing rabbitroutine.Consumer interface.
type Consumer struct {}
func (c *Consumer) Declare(ctx context.Context, ch *amqp.Channel) error {}
func (c *Consumer) Consume(ctx context.Context, ch *amqp.Channel) error {}

url := "amqp://guest:guest@127.0.0.1:5672/"

conn := rabbitroutine.NewConnector(rabbitroutine.Config{
    // How long to wait between reconnect
    Wait: 2 * time.Second,
})

ctx := context.Background()

go func() {
    err := conn.Dial(ctx, url)
    if err != nil {
    	log.Println(err)
    }
}()

consumer := &Consumer{}
go func() {
    err := conn.StartConsumer(ctx, consumer)
    if err != nil {
        log.Println(err)
    }
}()
```

[Full example demonstrates messages consuming](https://github.com/furdarius/rabbitroutine/blob/master/consumer_example_test.go)


### Publishing

For publishing [FireForgetPublisher](https://godoc.org/github.com/furdarius/rabbitroutine#FireForgetPublisher)
and [EnsurePublisher](https://godoc.org/github.com/furdarius/rabbitroutine#EnsurePublisher) implemented.
Both of them can be wrapped with [RetryPublisher](https://godoc.org/github.com/furdarius/rabbitroutine#RetryPublisher)
to repeat publishing on errors and mitigate short-term network problems.

Usage example:
```go
ctx := context.Background()

url := "amqp://guest:guest@127.0.0.1:5672/"

conn := rabbitroutine.NewConnector(rabbitroutine.Config{
    // How long wait between reconnect
    Wait: 2 * time.Second,
})

pool := rabbitroutine.NewPool(conn)
ensurePub := rabbitroutine.NewEnsurePublisher(pool)
pub := rabbitroutine.NewRetryPublisher(
    ensurePub,
    rabbitroutine.PublishMaxAttemptsSetup(16),
    rabbitroutine.PublishDelaySetup(rabbitroutine.LinearDelay(10*time.Millisecond)),
)

go conn.Dial(ctx, url)

err := pub.Publish(ctx, "myexch", "myqueue", amqp.Publishing{Body: []byte("message")})
if err != nil {
    log.Println("publish error:", err)
}

```

[Full example demonstrates messages publishing](https://github.com/furdarius/rabbitroutine/blob/master/publisher_example_test.go)

## Contributing

Pull requests are very much welcomed.  Create your pull request on a non-master
branch, make sure a test or example is included that covers your change and
your commits represent coherent changes that include a reason for the change.

To run the integration tests, make sure you have RabbitMQ running on any host,
export the environment variable `AMQP_URL=amqp://host/` and run `go test -tags
integration`. As example:
```
AMQP_URL=amqp://guest:guest@127.0.0.1:5672/ go test -v -race -cpu=1,2 -tags integration -timeout 5s
```

Use `gometalinter` to check code with linters:
```
gometalinter -t --vendor ./...
```

TravisCI will also run the integration tests and gometalinter.


CHANGELOG.md
# 0.5.0 | [Pull request](https://github.com/furdarius/rabbitroutine/pull/9)
- [EnsurePublisher.Publish](https://godoc.org/github.com/furdarius/rabbitroutine#EnsurePublisher.Publish) returns [ErrNotFound](https://godoc.org/github.com/furdarius/rabbitroutine#ErrNotFound) if RabbitMQ entity (e.g. exchange) doesn't exist.
- [EnsurePublisher.Publish](https://godoc.org/github.com/furdarius/rabbitroutine#EnsurePublisher.Publish) returns [ErrNoRoute](https://godoc.org/github.com/furdarius/rabbitroutine#ErrNoRoute) if message cannot be delivered to any queue.
- Functional options for configuring RetryPublisher added.
- NewRetryPublisherWithDelay deleted.

# 0.4.2 | [Pull request](https://github.com/furdarius/rabbitroutine/pull/8)
- Fix compile error on Windows `GOOS=windows GOARCH=386 go build` ([issue](https://github.com/furdarius/rabbitroutine/issues/7))

# 0.4.1 | [Pull request](https://github.com/furdarius/rabbitroutine/pull/6)
- Data race fixed on c.conn when connection error occurs.

# 0.4.0 | [Pull request](https://github.com/furdarius/rabbitroutine/pull/3)
- [LightningPool](https://godoc.org/github.com/furdarius/rabbitroutine#LightningPool) added.
- [FireForgetPublisher](https://godoc.org/github.com/furdarius/rabbitroutine#FireForgetPublisher) added.
- Possible deadlock on channel receiving from Dialed event listener fixed.
- [RetryPublisher](https://godoc.org/github.com/furdarius/rabbitroutine#RetryPublisher) accepts [Publisher](https://godoc.org/github.com/furdarius/rabbitroutine#Publisher) interface now.

# 0.3.1
- On error wait for c.cfg.Wait time before consumer restart.

# 0.3.0
- Rename Attempts to ReconnectAttempts in Connector config
- Make zero value of ReconnectAttempts equal infinity.
- Return error from ChannelKeeper Close method.

# 0.2.1
- On error once close amqp channel.

# 0.2.0
- Dial and DialConfig were added. DialConfig used to configure RabbitMQ connection settings.
- Config stores only reconnect options.
- Start replaced with Dial.



examples/publish/main.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/furdarius/rabbitroutine"
	"github.com/streadway/amqp"
)

func main() {
	ctx := context.Background()

	url := "amqp://guest:guest@127.0.0.1:5672/"

	conn := rabbitroutine.NewConnector(rabbitroutine.Config{
		// Max reconnect attempts
		ReconnectAttempts: 20,
		// How long wait between reconnect
		Wait: 2 * time.Second,
	})

	pool := rabbitroutine.NewPool(conn)
	ensurePub := rabbitroutine.NewEnsurePublisher(pool)
	pub := rabbitroutine.NewRetryPublisher(
		ensurePub,
		rabbitroutine.PublishMaxAttemptsSetup(16),
		rabbitroutine.PublishDelaySetup(rabbitroutine.LinearDelay(10*time.Millisecond)),
	)

	go func() {
		err := conn.Dial(ctx, url)
		if err != nil {
			log.Fatalf("failed to establish RabbitMQ connection: %v", err)
		}
	}()

	exchangeName := "myexchange"
	queueName := "myqueue"

	ch, err := conn.Channel(ctx)
	if err != nil {
		log.Fatalf("failed to create channel: %v", err)
	}

	err = ch.ExchangeDeclare(exchangeName, "direct", true, false, false, false, nil)
	if err != nil {
		log.Fatalf("failed to declare exchange: %v", err)
	}

	_, err = ch.QueueDeclare(queueName, true, false, false, false, nil)
	if err != nil {
		log.Fatalf("failed to declare queue: %v", err)
	}

	err = ch.QueueBind(queueName, queueName, exchangeName, false, nil)
	if err != nil {
		log.Fatalf("failed to bind queue: %v", err)
	}

	err = ch.Close()
	if err != nil {
		log.Printf("failed to close channel: %v", err)
	}

	for i := 0; i < 500000; i++ {
		timeoutCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)

		err := pub.Publish(timeoutCtx, exchangeName, queueName, amqp.Publishing{
			Body: []byte(fmt.Sprintf("message %d", i)),
		})
		if err != nil {
			log.Println("failed to publish:", err)
		}

		cancel()
	}
}
